internal/store/store.go

```diff
@@ -1,7 +1,41 @@
 package store
 
-var KV map[string]string
+import (
+	"context"
+	"errors"
+)
 
-func StoreInit() {
-	KV = make(map[string]string)
+// A Datastore is a storage mechanism for persisting urls mapping.
+type Datastore interface {
+	ReadURL(ctx context.Context, shortUrl string) (string, error)
+	WriteURL(ctx context.Context, shortUrl, url string) error
+}
+
+type memoryDatastore struct {
+	keyVal map[string]string
+}
+
+var _ Datastore = &memoryDatastore{}
+
+// ErrNotFound is thrown when the given short URL was not found in the datastore.
+var ErrNotFound = errors.New("short url not found in store")
+
+// NewMemoryDatastore instantiates a in-memory datastore.
+func NewMemoryDatastore() (Datastore, error) {
+	return &memoryDatastore{keyVal: make(map[string]string)}, nil
+}
+
+// ReadURL reads the datastore in an attempt to retreive the URL for the provided shortURL.
+func (ds *memoryDatastore) ReadURL(ctx context.Context, shortUrl string) (string, error) {
```

Comment:
why do we need the context?

Reply:
The `MemoryDatastore` does not really need this context but other datastore will need to know if a request has been cancelled to act on the interrupt.